### PR TITLE
Sylvia whittle/404 grainstats crash

### LINKED ISSUE #404: Grainstats crash
**Describe the bug**
A crash occurs while processing an AFM image which appears to originate from `grainstats.py`. The error is `ValueError: operands could not be broadcast together with shapes (0,) (2,) ` which is an issue with handling arrays. Check grainstats code to ensure that there are measures to check for bad array size combinations and also diagnose how they can arise. 

**To Reproduce**

Run TopoStats on the following file: https://drive.google.com/drive/folders/1_EUP6WyfsU5uSTO0eUXh1JUkxdsXDEab

Using the following config:

```yaml
base_dir: ./raphael/20220819/E4
output_dir: ./output/20220819/E4
warnings: ignore
cores: 4
quiet: false
file_ext: .spm
loading:
  channel: Height
filter:
  run: true # Options : true, false
  threshold_method: std_dev # Options : otsu, std_dev, absolute
  otsu_threshold_multiplier: 1.0
  threshold_std_dev:
    lower: 10.0
    upper: 1.0
  threshold_absolute:
    lower: -1.0
    upper: 1.0
  gaussian_size: 1.0121397464510862 # in px
  gaussian_mode: nearest
grains:
  run: true # Options : true, false
  # Thresholding by height
  threshold_method: std_dev # Options : std_dev, otsu, absolute
  otsu_threshold_multiplier: 1.0
  threshold_std_dev:
    lower: 10.0
    upper: 1.0
  threshold_absolute:
    lower: -1.0
    upper: 1.0
  direction: upper # Options: upper, lower, both (defines whether to look for grains above or below thresholds or both)
  # Thresholding by area
  absolute_smallest_grain_size: 100
  absolute_area_threshold:
    upper: [ 300, 100000 ] # above surface [Low, High] in nm^2 (also takes null)
    lower: [ null, null ] # below surface [Low, High] in nm^2 (also takes null)
grainstats:
  run: true # Options : true, false
  cropped_size: 40.0 # Length (in nm) of square cropped images (can take -1 for grain-sized box)
  save_cropped_grains: true # Options : true, false
dnatracing:
  run: true # Options : true, false
plotting:
  run: true # Options : true, false
  save_format: png # Options : see https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.savefig.html
  image_set: all  # Options : all, core
  zrange: [-4, 4]  # low and high height range for core images (can take [null, null])
  colorbar: true  # Options : true, false
  axes: true # Options : true, false (due to off being a bool when parsed)
  cmap: nanoscope # Options : nanoscope, afmhot
  histogram_log_axis: false # Options : true, false

```

**Expected behavior**

**Output**

```bash
❯ run_topostats --config ./my_config.yaml
[Wed, 14 Dec 2022 16:04:40] [INFO    ] [topostats] Configuration is valid.
[Wed, 14 Dec 2022 16:04:40] [INFO    ] [topostats] Plotting configuration is valid.
[Wed, 14 Dec 2022 16:04:40] [INFO    ] [topostats] Configuration file loaded from      : ./my_config.yaml
[Wed, 14 Dec 2022 16:04:40] [INFO    ] [topostats] Scanning for images in              : raphael/20220819/E4
[Wed, 14 Dec 2022 16:04:40] [INFO    ] [topostats] Output directory                    : output/20220819/E4
[Wed, 14 Dec 2022 16:04:40] [INFO    ] [topostats] Looking for images with extension   : .spm
[Wed, 14 Dec 2022 16:04:40] [INFO    ] [topostats] Images with extension .spm in raphael/20220819/E4 : 1
[Wed, 14 Dec 2022 16:04:40] [INFO    ] [topostats] Thresholding method (Filtering)     : std_dev
[Wed, 14 Dec 2022 16:04:40] [INFO    ] [topostats] Thresholding method (Grains)        : std_dev
[Wed, 14 Dec 2022 16:04:40] [INFO    ] [topostats] Extracting image from raphael/20220819/E4/E4-24h-nm.0_00006.spm
[Wed, 14 Dec 2022 16:04:40] [INFO    ] [topostats] Loading image from : raphael/20220819/E4/E4-24h-nm.0_00006.spm
[Wed, 14 Dec 2022 16:04:40] [INFO    ] [topostats] [E4-24h-nm.0_00006] : Loaded image from : raphael/20220819/E4/E4-24h-nm.0_00006.spm
[Wed, 14 Dec 2022 16:04:40] [INFO    ] [topostats] [E4-24h-nm.0_00006] : Extracted channel Height
[Wed, 14 Dec 2022 16:04:40] [INFO    ] [topostats] [E4-24h-nm.0_00006] : Pixel to nm scaling : 7.8125
Processing images from raphael/20220819/E4, results are under output/20220819/E4:   0%|       | 0/1 [00:00<?, ?it/s][Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats] image_path: raphael/20220819/E4/E4-24h-nm.0_00006
[Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats] image_path: raphael/20220819/E4/E4-24h-nm_0_00006
[Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats] Processing : E4-24h-nm_0_00006
[Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats] NO SUFFIX : 
[Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats] output_dir: output/20220819/E4
[Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats] _output_dir: output/20220819/E4/Processed
[Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats] processed path: output/20220819/E4/Processed
[Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats] folder path: raphael/20220819/E4
[Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats]  filter out path: output/20220819/E4/Processed/E4-24h-nm_0_00006/filters
[Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats]  grains out path: output/20220819/E4/Processed/E4-24h-nm_0_00006/grains
[Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Median flattening without mask
[Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Plane tilt removal without mask
[Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats] x-polyfit 1st order: [-0.04713146  6.01015864]
[Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats] y-polyfit 1st order: [ 1.15473115e-18 -1.93442339e-16]
[Wed, 14 Dec 2022 16:04:41] [INFO    ] [topostats] removing x plane tilt
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] removing y plane tilt
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Remove quadratic bow without mask
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] x polyfit 2nd order: [ 6.36937838e-06 -1.62419149e-03  6.07891608e+00]
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Masking (upper) Threshold: 6.356437682870245
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Masking (lower) Threshold: 2.5687851369480454
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Median flattening with mask
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Plane tilt removal with mask
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] x-polyfit 1st order: [ 2.53338788e-05 -3.03124756e-03]
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] y-polyfit 1st order: [ 2.41249252e-19 -6.02495788e-17]
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] removing x plane tilt
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] removing y plane tilt
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Remove quadratic bow with mask
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] x polyfit 2nd order: [ 6.19843233e-06 -1.58060024e-03  6.38808294e-02]
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Zero averaging background : -0.033168189250849044 nm
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Applying Gaussian filter (mode : nearest; Gaussian blur (px) : 1.0121397464510862).
[Wed, 14 Dec 2022 16:04:42] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Plotting Filtering Images
[Wed, 14 Dec 2022 16:04:46] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Image saved to : output/20220819/E4/Processed/E4-24h-nm_0_00006_processed
[Wed, 14 Dec 2022 16:04:46] [INFO    ] [topostats] [E4-24h-nm_0_00006] : *** Grain Finding ***
[Wed, 14 Dec 2022 16:04:46] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Thresholding method (grains) : std_dev
[Wed, 14 Dec 2022 16:04:46] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing upper threshold (0.2577153403696904)
[Wed, 14 Dec 2022 16:04:46] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Masking (upper) Threshold: 0.2577153403696904
[Wed, 14 Dec 2022 16:04:46] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Labelling Regions
[Wed, 14 Dec 2022 16:04:46] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Tidying borders
[Wed, 14 Dec 2022 16:04:46] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Labelling Regions
[Wed, 14 Dec 2022 16:04:46] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Region properties calculated (upper)
[Wed, 14 Dec 2022 16:04:46] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Coloured regions
[Wed, 14 Dec 2022 16:04:46] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Extracted bounding boxes (upper)
[Wed, 14 Dec 2022 16:04:46] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Plotting Grain Finding Images
[Wed, 14 Dec 2022 16:04:47] [INFO    ] [topostats] [E4-24h-nm_0_00006_upper] : Image saved to : output/20220819/E4/Processed/E4-24h-nm_0_00006_upper_processed_masked
[Wed, 14 Dec 2022 16:04:47] [INFO    ] [topostats] [E4-24h-nm_0_00006] : *** Grain Statistics ***
[Wed, 14 Dec 2022 16:04:47] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 0
[Wed, 14 Dec 2022 16:04:47] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 1
[Wed, 14 Dec 2022 16:04:47] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 2
[Wed, 14 Dec 2022 16:04:48] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 3
[Wed, 14 Dec 2022 16:04:48] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 4
[Wed, 14 Dec 2022 16:04:48] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 5
[Wed, 14 Dec 2022 16:04:49] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 6
[Wed, 14 Dec 2022 16:04:49] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 7
[Wed, 14 Dec 2022 16:04:49] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 8
[Wed, 14 Dec 2022 16:04:49] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 9
[Wed, 14 Dec 2022 16:04:50] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 10
[Wed, 14 Dec 2022 16:04:50] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 11
[Wed, 14 Dec 2022 16:04:50] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 12
[Wed, 14 Dec 2022 16:04:50] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 13
[Wed, 14 Dec 2022 16:04:51] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 14
[Wed, 14 Dec 2022 16:04:51] [INFO    ] [topostats] [E4-24h-nm_0_00006] : Processing grain: 15
Processing images from raphael/20220819/E4, results are under output/20220819/E4:   0%|       | 0/1 [00:11<?, ?it/s]
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/Users/sylvi/opt/anaconda3/envs/scars/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/Users/sylvi/Documents/GitKraken/TopoStats/topostats/run_topostats.py", line 329, in process_scan
    ).calculate_stats()
  File "/Users/sylvi/Documents/GitKraken/TopoStats/topostats/grainstats.py", line 218, in calculate_stats
    radius_stats = self.calculate_radius_stats(edges, points)
  File "/Users/sylvi/Documents/GitKraken/TopoStats/topostats/grainstats.py", line 368, in calculate_radius_stats
    displacements = self._calculate_displacement(edges, centroid)
  File "/Users/sylvi/Documents/GitKraken/TopoStats/topostats/grainstats.py", line 400, in _calculate_displacement
    return np.array(edges) - centroid
ValueError: operands could not be broadcast together with shapes (0,) (2,) 
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/sylvi/opt/anaconda3/envs/scars/bin/run_topostats", line 8, in <module>
    sys.exit(main())
  File "/Users/sylvi/Documents/GitKraken/TopoStats/topostats/run_topostats.py", line 459, in main
    for img, result in pool.imap_unordered(
  File "/Users/sylvi/opt/anaconda3/envs/scars/lib/python3.10/multiprocessing/pool.py", line 873, in next
    raise value
ValueError: operands could not be broadcast together with shapes (0,) (2,) 
```

**(Not required if using Docker) Your computer configuration (please complete the following information):**
 - OS: MacOS
 - topostats version: 2.0.0
 - Python version: 3.10

### PR DESCRIPTION
**Pull request by SylviaWhittle**
Tiny PR! 

Closes #404 

This PR aims to fix bug #404, by both preventing tiny grains (less than 5x5) being processed in grainstats, and also to improve the edge finding algorithm so that if in future we want to handle grains less than 5x5, `calculate_edges` will be able to handle this. (Note that other functions may still have issue with this). So in effect, this *should* be both prevention and cure for the bug. 

I have verified that this works locally on my machine, however for you to verify that it fixes the bug, run latest `main` TopoStats on this file: https://drive.google.com/drive/folders/1jAkiS0KKLBmOw8rKLZ-nrgy4MIhfyoJT (let me know if you need permissions to access), with this config:

```yaml
base_dir: ./
output_dir: output
warnings: ignore
cores: 2
quiet: false
file_ext: .spm
loading:
  channel: Height
filter:
  run: true
  threshold_method: std_dev
  otsu_threshold_multiplier: 1.0
  threshold_std_dev:
    lower: 10.0
    upper: 1.3
  threshold_absolute:
    lower: -1.0
    upper: 1.0
  gaussian_size: 1.0121397464510862
  gaussian_mode: nearest
  remove_scars:
    run: true
    removal_iterations: 2
    threshold_low: 0.25
    threshold_high: 0.666
    max_scar_width: 4
    min_scar_length: 16
grains:
  run: true
  threshold_method: std_dev
  otsu_threshold_multiplier: 1.0
  threshold_std_dev:
    lower: 10.0
    upper: 1.3
  threshold_absolute:
    lower: -1.0
    upper: 1.0
  direction: upper
  smallest_grain_size_nm2: 10
  absolute_area_threshold:
    upper:
    - 10
    - 100000
    lower:
    - 
    - 
grainstats:
  run: true
  cropped_size: 40.0
dnatracing:
  run: true
plotting:
  run: true
  save_format: png
  image_set: core
  zrange: [-3, 4]
  colorbar: true
  axes: true
  cmap: nanoscope
  mask_cmap: blu
  histogram_log_axis: false
  histogram_bins: 200

```

For me, this fixes the issue, allowing another bug (#402 ) to occur, but that is for me to fix next 😄 

Snippet:
```bash
[Wed, 08 Feb 2023 12:20:39] [INFO    ] [topostats] [E4-24h-nm.0_00006] : *** DNA Tracing ***
[Wed, 08 Feb 2023 12:20:39] [INFO    ] [topostats] [E4-24h-nm.0_00006] : Gaussian filter applied.
Processing images from /Users/sylvi/topo_data/404-grainstats-crash, results are 
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/Users/sylvi/Documents/TopoStats/topostats/run_topostats.py", line 334, in process_scan
    dna_traces[direction].trace_dna()
  File "/Users/sylvi/Documents/TopoStats/topostats/tracing/dnatracing.py", line 91, in trace_dna
    self.get_splined_traces()
  File "/Users/sylvi/Documents/TopoStats/topostats/tracing/dnatracing.py", line 415, in get_splined_traces
    while nbr / step_size < 4:
ZeroDivisionError: division by zero
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/miniforge/base/bin/run_topostats", line 8, in <module>
    sys.exit(main())
  File "/Users/sylvi/Documents/TopoStats/topostats/run_topostats.py", line 442, in main
    for img, result in pool.imap_unordered(
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/multiprocessing/pool.py", line 873, in next
    raise value
ZeroDivisionError: division by zero
```

I have decided not to make this 5x5 option configurable, as I don't think this is a user-facing option, since it is inherently tied to the processing limitations and allowing users to ignore or change this check could break things for them. However, that being said, please do say if you think this is unwise.

### CODE REVIEW

**Comment by SylviaWhittle** [comment]
I have updated the regtests, these updates are minor and accounted for. 

The previously used method for detecting edges of a grain, using `canny` filtering, produced smoothed edges, which are not completely accurate, but good approximations. The updated method using erosion reliably produces exact edges. Here is an example:

image

LHS: Grain mask
Mid: Old canny edge-finding
RHS: New erosion edge-finding

File: topostats/grainstats.py [hunk 1]
```diff
@@ -173,6 +173,15 @@ def calculate_stats(self) -> Dict:
         # There are multiple entries for each grain.
         for index, region in enumerate(region_properties):
             LOGGER.info(f"[{self.image_name}] : Processing grain: {index}")
+
+            # Skip grain if too small to calculate stats for
+            LOGGER.debug(f"[{self.image_name}] : Grain size: {region.image.size}")
+            if min(region.image.shape) < 5:
```
Comment (ns-rse) [inline topostats/grainstats.py:179]: Good to see you caught my dumb mistake!

**Review by ns-rse — COMMENTED** [review]
Looks ok code wise to me.

I can't really comment on the difference between [`canny`](https://scikit-image.org/docs/dev/auto_examples/edges/plot_canny.html) v [`erosion`](https://scikit-image.org/docs/stable/api/skimage.morphology.html#skimage.morphology.erosion) for defining boundaries though (I see there is also [`binary_erosion](https://scikit-image.org/docs/stable/api/skimage.morphology.html#skimage.morphology.binary_erosion) which does the same but faster, perhaps worth considering if at this point erosion is done on a binary image :shrug: ). `canny` used a fixed value of `sigma=3` so perhaps this might have been something that was configurable by users and might have made a difference?

For example the `canny` method gives smooth edges which might not be accurate, but then @Jean-Du has introduced a splining method to DNA Tracing which is, based on my crude understanding, also a method of smoothing curves. What is the preferred method _overall_? (That is a rhetorical question though and I actually think giving people options to choose what works best for them is preferable).

Approving but perhaps we need to give some thought as to what we want to use/make available. We could perhaps consider making the method (`canny` v `erosion` v `binary_erosion`) configurable so that users are able to choose what is best for them.

**Comment by SylviaWhittle** [comment]
Thank you @ns-rse . Those are some excellent points. 

I will field the options to the experimentalists in a meeting tomorrow morning to see what would be preferable and if they would like the option of `canny` edge detection. I would guess that accuracy at the grainstats stage is imperative given that small changes in statistics like `min_feret` are used in research to determine effectivenesses of different therapeutics.

You are right about `binary_erosion`, I have updated that accordingly.

Perfectly happy to wait for #455 before merging 👍

**Comment by MaxGamill-Sheffield** [comment]
I was just thinking about this - is there a reason these tiny 5x5 grains aren't removed by the "absolute smallest grain size". And should this then be a value in pixels of this limit, to ensure the edge detection works?

I also prefer the mask-edge accuracy of the `erosion` method, although both `erosion` and `canny` methods require accurate molecule masks to produce accurate statistics, and `canny` may help to alleviate this (testing needed).

**Comment by SylviaWhittle** [comment]
@MaxGamill-Sheffield So 5x5 grains are not usually encountered, unless the user sets the size thresholding values to really low (like I accidentally did at one point while processing Raphael's data, which is how this bug was found).

I don't think this should be a size defined in nanometres, since it's a limitation directly tied to pixel size (`canny` cannot process less than 5x5 correctly), not the real-world length scale. If it was tied to nanometres, then images with different `pixel_to_nm_scaling_factor`s might trigger this issue again.

Given the helpful feedback, I think I might keep both methods in, and have a default of `erosion`. What do you both think of this?

**Review by ns-rse — CHANGES_REQUESTED** [review]
One very minor modification to `docs/configuration.md` but otherwise good to go. :+1:

**Review by ns-rse — APPROVED** [review]
On :fire: ! More than happy to approve.

**Pull request merged**